LEGO1/mxpalette.cpp

```diff
@@ -1,8 +1,47 @@
 #include "mxpalette.h"
 
+// OFFSET: LEGO1 0x100bee30
+MxPalette::MxPalette()
+{
+  this->m_overrideSkyColor = FALSE;
+  this->m_attached = NULL;
+  GetDefaultPalette(this->m_entries);
+  this->m_skyColor = this->m_entries[0x8d];
+  // FIXME: Incomplete
+}
+
+// OFFSET: LEGO1 100bef90
+MxPalette::~MxPalette()
+{
+  delete m_attached;  // yes this matches more
+}
+
 // OFFSET: LEGO1 0x100bf150
 MxResult MxPalette::GetEntries(LPPALETTEENTRY p_entries)
 {
   memcpy(p_entries, this->m_entries, sizeof(this->m_entries));
   return SUCCESS;
+}
+
+// OFFSET: LEGO1 0x100bf420
+void MxPalette::GetDefaultPalette(LPPALETTEENTRY p_entries)
+{
+  HDC hdc = GetDC((HWND) NULL);
+  int rasterCaps = GetDeviceCaps(hdc, RASTERCAPS);
+  if ((rasterCaps & RC_PALETTE) != 0) {
+    int paletteSize = GetDeviceCaps(hdc, SIZEPALETTE);
+    if (paletteSize == 256) {
+      GetSystemPaletteEntries(hdc, 0, 256, p_entries);
+      p_entries += 10;
+      memcpy(p_entries, this->m_entries, sizeof(this->m_entries));
+    }
+  }
+  
+  ReleaseDC((HWND) NULL, hdc);
```

Comment:
I got this, but it doesn't make total sense to me.
`g_defaultPalette` is at `10102188`.

```suggestion
  HDC hdc = GetDC((HWND) NULL);
  PALETTEENTRY *src;

  if ((rasterCaps & RC_PALETTE) != 0 && GetDeviceCaps(hdc, SIZEPALETTE) == 256) {
    GetSystemPaletteEntries(hdc, 0, 256, p_entries);
    count = 256 - 2 * 10;
    src = &g_defaultPalette[10];
    p_entries += 10;
  } else {
    src = g_defaultPalette;
    count = 256;
  }
  memcpy(dest, p_entries, count * sizeof(PALETTEENTRY));
  ReleaseDC((HWND) NULL, hdc);
```

Reply:
I kinda understand what you did here, I'll take your word for it

Reply:
ehh the build isn't working with this, can you PR my branch ItzSwirlz:mxpalette

Reply:
This code was written untested in the browser. You need to fix the errors...

Reply:
sorry, i'm tired, it's been a long day. I'll do it

Reply:
![image](https://github.com/isledecomp/isle/assets/24401303/7702cba5-c042-4d58-84c6-dcfd11213564)
aight, so this matches 62.07%.... ill need to look back at stuff later, Clone is the only other function I'll try today

Reply:
`src` is `PALETTEENTRY*` (a.k.a. `LPPALETTEENTRY`), not `IDirectDrawPalette*`.

Some ancient history might be in place here (both for your understanding and to add a comment to the code). The reason it skips the first ten palette entries is that they are the "system palette", i.e. the colors Windows itself uses for drawing things like windows, buttons and so on. That also applies the the last ten entries here.

Reply:
Thanks! That's very useful.